src/hooks/useZapExecute.ts

```diff
@@ -0,0 +1,165 @@
+import { t } from "@lingui/macro";
+import { BigNumber, ContractReceipt, ethers } from "ethers";
+import { useMutation, useQueryClient } from "react-query";
+import { useDispatch } from "react-redux";
+import { NetworkId } from "src/constants";
+import { GOHM_ADDRESSES } from "src/constants/addresses";
+import { trackGAEvent } from "src/helpers/analytics";
+import { addresses } from "src/networkDetails";
+import { error, info } from "src/slices/MessagesSlice";
+
+import { SOHM_ADDRESSES } from "../constants/addresses";
+import { DecimalBigNumber } from "../helpers/DecimalBigNumber/DecimalBigNumber";
+import { Environment } from "../helpers/environment/Environment/Environment";
+import { Zap__factory } from "../typechain/factories/Zap__factory";
+import { balanceQueryKey } from "./useBalance";
+import { zapTokenBalancesKey } from "./useZapTokenBalances";
+import { useWeb3Context } from "./web3Context";
+
+interface ZapTransactionResponse {
+  to: string;
+  data: string;
+  estimatedGas: string;
+  buyAmount: string;
+}
+
+interface IUADataZap {
+  address: string;
+  value: string;
+  token: string;
+  type: string;
+  slippage: string;
+  approved: boolean;
+}
+
+interface ZapExecuteOptions {
+  slippage: string;
+  sellAmount: BigNumber;
+  tokenAddress: string;
+  minimumAmount: string;
+  gOHM: boolean;
+}
+
+export const useZapExecute = () => {
+  const dispatch = useDispatch();
+  const client = useQueryClient();
+  const { address, provider, networkId } = useWeb3Context();
+  const signer = provider.getSigner();
+  // We only operate on Ethereum mainnet for the moment, so we can use a static contract
+  const contract = Zap__factory.connect(addresses[networkId].ZAP, signer);
+
+  return useMutation<ContractReceipt, Error, ZapExecuteOptions>(
+    /**
+     * Ideally the parameters to this async function should be the slippage, etc.
+     * However the `mutationFn` parameter to `useMutation` accepts a function with
+     * one parameter only, and it wasn't working when passing an object with a defined interface.
+     * So the parameters are moved up a level.
+     */
+    async ({ slippage, sellAmount, tokenAddress, minimumAmount, gOHM }) => {
+      // TODO throw or dispatch error?
```

Comment:
@sam-potter do you have a suggestion on whether to throw or dispatch errors during the mutation?